File: jira/frontend-reactjs/src/types/task.ts
// ============================================================================
// Task Types
// ============================================================================

import type { Project } from "./project";
import type { Comment } from "./comment";
import type { ActivityLog } from "./activity";

// ── Enums / Union types ─────────────────────────────────────────────────────

export type TaskStatus = "todo" | "in-progress" | "review" | "done";

export type TaskPriority = "low" | "medium" | "high" | "urgent";

export type IssueType = "story" | "task" | "bug" | "epic" | "subtask";

export type LinkType = "blocks" | "is-blocked-by" | "relates-to" | "duplicates";

// ── Linked Issue ────────────────────────────────────────────────────────────

export interface LinkedIssue {
  id: string;
  link_type: LinkType;
  task: TaskSummary;
}

// ── Entity ──────────────────────────────────────────────────────────────────

export interface Task {
  id: string;
  project_id: string;
  issue_key: string;
  issue_type: IssueType;
  title: string;
  description?: string;
  status: TaskStatus;
  priority: TaskPriority;
  reporter: string;
  assignee: string;
  due_date: string | null;
  sprint_id: string | null;
  epic_id: string | null;
  story_points: number | null;
  labels: string[];
  parent_id: string | null;
  subtasks?: TaskSummary[];
  linked_issues?: LinkedIssue[];
  project?: Project;
  comments?: Comment[];
  activities?: ActivityLog[];
  created_at: string;
  updated_at: string;
}

/** Compact task reference (board cards, search results, linked issues) */
export interface TaskSummary {
  id: string;
  issue_key: string;
  issue_type: IssueType;
  title: string;
  status: TaskStatus;
  priority: TaskPriority;
  assignee: string;
}

// ── Form / Request payloads ─────────────────────────────────────────────────

export interface TaskFormData {
  title: string;
  description?: string;
  issue_type: IssueType;
  status: TaskStatus;
  priority: TaskPriority;
  assignee?: string;
  reporter?: string;
  due_date?: string;
  sprint_id?: string;
  epic_id?: string;
  parent_id?: string;
  story_points?: number;
  labels?: string[];
}

export type CreateTaskRequest = TaskFormData;

export type UpdateTaskRequest = Partial<TaskFormData>;

// ── Query helpers ───────────────────────────────────────────────────────────

export interface TaskQueryParams {
  page?: number;
  page_size?: number;
  search?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  issue_type?: IssueType;
  assignee?: string;
  reporter?: string;
  sprint_id?: string;
  sort_by?: string;
  sort_direction?: "asc" | "desc";
}

File: project-hub/backend/app/schemas/task.py
from pydantic import BaseModel


class TaskCreate(BaseModel):
    title: str
    description: str = ""
    issue_type: str = "task"
    status: str = "todo"
    priority: str = "medium"
    assignee: str | None = None
    reporter: str | None = None
    due_date: str | None = None
    sprint_id: str | None = None
    epic_id: str | None = None
    parent_id: str | None = None
    story_points: float | None = None
    labels: list[str] = []


class TaskUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    issue_type: str | None = None
    status: str | None = None
    priority: str | None = None
    assignee: str | None = None
    reporter: str | None = None
    due_date: str | None = None
    sprint_id: str | None = None
    epic_id: str | None = None
    parent_id: str | None = None
    story_points: float | None = None
    labels: list[str] | None = None


class TaskResponse(BaseModel):
    id: str
    project_id: str
    issue_key: str
    title: str
    description: str
    issue_type: str
    status: str
    priority: str
    assignee: str | None = None
    reporter: str | None = None
    due_date: str | None = None
    sprint_id: str | None = None
    epic_id: str | None = None
    parent_id: str | None = None
    story_points: float | None = None
    labels: list[str] = []
    created_at: str
    updated_at: str

    model_config = {"from_attributes": True}


class TaskSummary(BaseModel):
    id: str
    issue_key: str
    title: str
    status: str
    priority: str
    assignee: str | None = None

    model_config = {"from_attributes": True}


class LinkedIssueResponse(BaseModel):
    id: str
    link_type: str
    task: TaskSummary

    model_config = {"from_attributes": True}


class CreateSubtaskRequest(BaseModel):
    title: str
    assignee: str | None = None


class CreateLinkRequest(BaseModel):
    target_task_id: str
    link_type: str


class BulkUpdateRequest(BaseModel):
    task_ids: list[str]
    status: str | None = None
    priority: str | None = None
    assignee: str | None = None


class BulkDeleteRequest(BaseModel):
    task_ids: list[str]
